File: src/cavebot/navigation/atlas.py
"""Minimap atlas — loads PNG tiles into an indexed structure."""

from __future__ import annotations

import re
from pathlib import Path

import numpy as np
from PIL import Image

_COLOR_RE = re.compile(r"Minimap_Color_(\d+)_(\d+)_(\d+)\.png")
_COST_RE = re.compile(r"Minimap_WaypointCost_(\d+)_(\d+)_(\d+)\.png")
CHUNK_SIZE = 256


class Atlas:
    def __init__(self, minimap_dir: str | Path):
        self.minimap_dir = Path(minimap_dir)
        self.color_chunks: dict[tuple[int, int, int], np.ndarray] = {}
        self.cost_chunks: dict[tuple[int, int, int], np.ndarray] = {}
        self.z_levels: set[int] = set()

    def load(self) -> None:
        """Scan minimap directory and load all tiles into memory."""
        for fpath in self.minimap_dir.iterdir():
            m = _COLOR_RE.match(fpath.name)
            if m:
                cx, cy, cz = int(m.group(1)), int(m.group(2)), int(m.group(3))
                img = Image.open(fpath).convert("RGB")
                self.color_chunks[(cx, cy, cz)] = np.array(img)
                self.z_levels.add(cz)
                continue
            m = _COST_RE.match(fpath.name)
            if m:
                cx, cy, cz = int(m.group(1)), int(m.group(2)), int(m.group(3))
                img = Image.open(fpath)
                self.cost_chunks[(cx, cy, cz)] = np.array(img)
                self.z_levels.add(cz)

    def get_color_tile(self, chunk_x: int, chunk_y: int, z: int) -> np.ndarray | None:
        return self.color_chunks.get((chunk_x, chunk_y, z))

    def get_cost_tile(self, chunk_x: int, chunk_y: int, z: int) -> np.ndarray | None:
        return self.cost_chunks.get((chunk_x, chunk_y, z))

    def get_cost_at(self, world_x: int, world_y: int, z: int) -> int:
        """Get walk cost at an absolute world coordinate. Returns 255 if missing."""
        chunk_x = (world_x // CHUNK_SIZE) * CHUNK_SIZE
        chunk_y = (world_y // CHUNK_SIZE) * CHUNK_SIZE
        tile = self.cost_chunks.get((chunk_x, chunk_y, z))
        if tile is None:
            return 255
        px = world_x - chunk_x
        py = world_y - chunk_y
        if 0 <= px < CHUNK_SIZE and 0 <= py < CHUNK_SIZE:
            return int(tile[py, px])
        return 255

    def chunk_keys_for_z(self, z: int) -> list[tuple[int, int, int]]:
        return [k for k in self.color_chunks if k[2] == z]

    def world_bounds(self, z: int) -> dict[str, int]:
        keys = self.chunk_keys_for_z(z)
        if not keys:
            return {"min_x": 0, "min_y": 0, "max_x": 0, "max_y": 0}
        min_x = min(k[0] for k in keys)
        min_y = min(k[1] for k in keys)
        max_x = max(k[0] for k in keys) + CHUNK_SIZE
        max_y = max(k[1] for k in keys) + CHUNK_SIZE
        return {"min_x": min_x, "min_y": min_y, "max_x": max_x, "max_y": max_y}

File: go-cavebot/internal/navigation/atlas.go
package navigation

import (
	"image"
	"image/color"
	"image/png"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
)

const ChunkSize = 256

type ChunkKey struct {
	X, Y, Z int
}

type Bounds struct {
	MinX int `json:"min_x"`
	MinY int `json:"min_y"`
	MaxX int `json:"max_x"`
	MaxY int `json:"max_y"`
}

var (
	colorRe = regexp.MustCompile(`Minimap_Color_(\d+)_(\d+)_(\d+)\.png`)
	costRe  = regexp.MustCompile(`Minimap_WaypointCost_(\d+)_(\d+)_(\d+)\.png`)
)

type Atlas struct {
	dir         string
	ColorChunks map[ChunkKey]*image.RGBA
	CostChunks  map[ChunkKey]*image.Gray
	ZLevels     map[int]bool
}

func NewAtlas(dir string) *Atlas {
	return &Atlas{
		dir:         dir,
		ColorChunks: make(map[ChunkKey]*image.RGBA),
		CostChunks:  make(map[ChunkKey]*image.Gray),
		ZLevels:     make(map[int]bool),
	}
}

func (a *Atlas) Load() error {
	entries, err := os.ReadDir(a.dir)
	if err != nil {
		return err
	}
	for _, entry := range entries {
		name := entry.Name()

		if m := colorRe.FindStringSubmatch(name); m != nil {
			cx, _ := strconv.Atoi(m[1])
			cy, _ := strconv.Atoi(m[2])
			cz, _ := strconv.Atoi(m[3])
			key := ChunkKey{cx, cy, cz}
			img, err := loadPNG(filepath.Join(a.dir, name))
			if err != nil {
				continue
			}
			a.ColorChunks[key] = toRGBA(img)
			a.ZLevels[cz] = true
			continue
		}

		if m := costRe.FindStringSubmatch(name); m != nil {
			cx, _ := strconv.Atoi(m[1])
			cy, _ := strconv.Atoi(m[2])
			cz, _ := strconv.Atoi(m[3])
			key := ChunkKey{cx, cy, cz}
			img, err := loadPNG(filepath.Join(a.dir, name))
			if err != nil {
				continue
			}
			a.CostChunks[key] = toGray(img)
			a.ZLevels[cz] = true
		}
	}
	return nil
}

func loadPNG(path string) (image.Image, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	return png.Decode(f)
}

func toRGBA(img image.Image) *image.RGBA {
	if rgba, ok := img.(*image.RGBA); ok {
		return rgba
	}
	bounds := img.Bounds()
	rgba := image.NewRGBA(bounds)
	for y := bounds.Min.Y; y < bounds.Max.Y; y++ {
		for x := bounds.Min.X; x < bounds.Max.X; x++ {
			rgba.Set(x, y, img.At(x, y))
		}
	}
	return rgba
}

func toGray(img image.Image) *image.Gray {
	if gray, ok := img.(*image.Gray); ok {
		return gray
	}
	bounds := img.Bounds()
	gray := image.NewGray(bounds)
	for y := bounds.Min.Y; y < bounds.Max.Y; y++ {
		for x := bounds.Min.X; x < bounds.Max.X; x++ {
			r, g, b, _ := img.At(x, y).RGBA()
			// Use luminance for grayscale conversion
			lum := (r*299 + g*587 + b*114) / 1000
			gray.SetGray(x, y, color.Gray{Y: uint8(lum >> 8)})
		}
	}
	return gray
}

func (a *Atlas) GetColorTile(chunkX, chunkY, z int) (*image.RGBA, bool) {
	tile, ok := a.ColorChunks[ChunkKey{chunkX, chunkY, z}]
	return tile, ok
}

func (a *Atlas) GetCostAt(worldX, worldY, z int) int {
	chunkX := (worldX / ChunkSize) * ChunkSize
	chunkY := (worldY / ChunkSize) * ChunkSize
	tile, ok := a.CostChunks[ChunkKey{chunkX, chunkY, z}]
	if !ok {
		return 255
	}
	px := worldX - chunkX
	py := worldY - chunkY
	if px < 0 || px >= ChunkSize || py < 0 || py >= ChunkSize {
		return 255
	}
	return int(tile.GrayAt(px, py).Y)
}

func (a *Atlas) ChunkKeysForZ(z int) []ChunkKey {
	var keys []ChunkKey
	for k := range a.ColorChunks {
		if k.Z == z {
			keys = append(keys, k)
		}
	}
	return keys
}

func (a *Atlas) WorldBounds(z int) Bounds {
	keys := a.ChunkKeysForZ(z)
	if len(keys) == 0 {
		return Bounds{}
	}
	minX, minY := keys[0].X, keys[0].Y
	maxX, maxY := keys[0].X, keys[0].Y
	for _, k := range keys[1:] {
		if k.X < minX {
			minX = k.X
		}
		if k.Y < minY {
			minY = k.Y
		}
		if k.X > maxX {
			maxX = k.X
		}
		if k.Y > maxY {
			maxY = k.Y
		}
	}
	return Bounds{
		MinX: minX,
		MinY: minY,
		MaxX: maxX + ChunkSize,
		MaxY: maxY + ChunkSize,
	}
}
